src/libfuncs/array.rs

```diff
@@ -371,720 +423,1115 @@ pub fn build_get<'ctx, 'this>(
+                        ));
+                    }
+                    _ => block.memcpy(context, location, data_ptr, value_ptr, value_size),
+                }
 
-    let ptr_ty = crate::ffi::get_struct_field_type_at(&array_ty, 0);
-    let len_ty = crate::ffi::get_struct_field_type_at(&array_ty, 1);
+                let array_ptr = {
+                    let array_len_bytes =
+                        block.append_op_result(arith::subi(array_len, k1, location))?;
+                    let array_len_bytes = block.append_op_result(arith::extui(
+                        array_len_bytes,
+                        IntegerType::new(context, 64).into(),
+                        location,
+                    ))?;
+                    let array_len_bytes = block.append_op_result(arith::muli(
+                        array_len_bytes,
+                        elem_stride,
+                        location,
+                    ))?;
+                    let array_len_bytes = block.append_op_result(arith::addi(
+                        array_len_bytes,
+                        block.const_int(
+                            context,
+                            location,
+                            calc_refcount_offset(elem_layout),
+                            64,
+                        )?,
+                        location,
+                    ))?;
+
+                    let clone_ptr = block.append_op_result(llvm::zero(ptr_ty, location))?;
+                    let clone_ptr = block.append_op_result(ReallocBindingsMeta::realloc(
+                        context,
+                        clone_ptr,
+                        array_len_bytes,
+                        location,
+                    )?)?;
+                    block.store(context, location, clone_ptr, k1)?;
+
+                    let clone_ptr = block.gep(
+                        context,
+                        location,
+                        clone_ptr,
+                        &[GepIndex::Const(calc_refcount_offset(elem_layout) as i32)],
+                        IntegerType::new(context, 8).into(),
+                    )?;
+
+                    let data_ptr = if REVERSE {
+                        data_ptr
+                    } else {
+                        let offset = block.append_op_result(arith::extui(
+                            extract_len_value,
+                            IntegerType::new(context, 64).into(),
+                            location,
+                        ))?;
+                        let offset =
+                            block.append_op_result(arith::muli(offset, elem_stride, location))?;
 
-    let value = entry.arg(0)?;
+                        block.gep(
+                            context,
+                            location,
+                            array_ptr,
+                            &[GepIndex::Value(offset)],
+                            IntegerType::new(context, 8).into(),
+                        )?
+                    };
```

Comment:
- `array_ptr`: The pointer that is stored in the `!llvm.struct<(ptr, i32, i32, i32)>`. It may or may not point to the data.
  - `data_ptr`: It's `array_ptr + elem_stride * array_start`. It always points to the start of the data. It'll be equivalent to `array_ptr` when `array_start` is zero.

When popping in reverse we start copying data from where the data start and skip the popped elements, thus the start pointer is `data_ptr`. The offset in the cloned array will always be zero (since we're copying the data anyways we'll also defragment it.

Reply:
Aah, I get it now. Thanks